[Pages] Add docs for configuring Pages Functions projects via Wrangler. DO NOT MERGE. (#13630)
cloudflare/cloudflare-docs: content/pages/functions/wrangler-configuration.md

- this file becomes the [source of truth](/pages/functions/wrangler-configuration/#source-of-truth) when used, meaning that you **can not edit the same fields in the dashboard** once you are using this file

## Configure environments

With `wrangler.toml` you can quickly set configuration across your local environment, preview deployments, and production.

### Local development

`wrangler.toml` works locally when using `wrangler pages dev`. This means that you can test out configuration changes quickly without a need to login to the Cloudflare dashboard. Refer to the following config file for an example:

```toml
---
filename: wrangler.toml
---
name = "my-pages-app"
pages_build_output_dir = "./dist"
compatibility_date = "2023-10-12"
compatibility_flags = ["nodejs_compat"]

[[kv_namespaces]]
binding = "KV"
id = "<NAMESPACE_ID>"
```

This config file adds the `nodejs_compat` compatibility flag and a KV namespace binding to your project. Running `wrangler pages dev` in a directory with this config file will apply the `nodejs_compat` compatibility flag locally, and expose the `KV` binding in your Pages Function code at `context.env.KV`.

{{<Aside type="note">}}

> For a full list of configuration keys refer to [inheritable keys](#inheritable-keys) and [non-inheritable keys](#non-inheritable-keys).

Reviewer: ```suggestion
For a full list of configuration keys, refer to [inheritable keys](#inheritable-keys) and [non-inheritable keys](#non-inheritable-keys).
```

{{</Aside>}}

### Production and Preview Deployments

> Once you are ready to deploy your project, you can set the configuration for production and preview deployments by creating a new deployment containing a `wrangler.toml` file.

Reviewer: Is this necessary because of wrangler.toml? These things are set in the dashboard normally, no?

Author: Either way, you have to make a new deployment before config values take effect. Pages behaves differently than Workers with this.

{{<Aside type="note">}}

For the following commands, it is important to remember the branch that you set as your [production branch](/pages/configuration/branch-build-controls/#production-branch-control) as well as your [preview branch settings](/pages/configuration/branch-build-controls/#preview-branch-control).

{{</Aside>}}

To use the configuration example above for production configuration, you can run:

```sh

> $ npx wrangler pages deploy --branch <PRODUCTION BRANCH>

Reviewer: This conflicts with https://f7a607c8.cloudflare-docs-7ou.pages.dev/workers/wrangler/commands/#deploy-1

Is build output dir required or not?

Author: I'm not sure what you're referring to here.

```

To deploy the configuration for preview deployments, you can run the same command as above, and change the `--branch` name to any branch besides the one you set as the production branch.

This will set the configuration for all preview deployments, not just the deployments from a specific branch. Pages does not currently support branch-based configuration.

{{<Aside type="note">}}

> `--branch` is optional. If you omit it, Wrangler will infer the branch you are on from the repo you are currently in and implicitly add it to the command.

Reviewer: ```suggestion
`--branch` is optional. If you omit it, Wrangler will infer the branch you are on from the repository you are currently in and implicitly add it to the command.
```

{{</Aside>}}

### Environment-specific overrides

There are times that you might want to use different configuration across local, preview, and production. You can override configuration for production and preview deployments by using `[env.production]` or `[env.preview]`.

{{<Aside type="note">}}

> Unlike [Workers Environments](/workers/wrangler/configuration/#environments), `production` and `preview` are the only two options available via `[env.name]`.

Reviewer: ```suggestion
Unlike [Workers environments](/workers/wrangler/configuration/#environments), `production` and `preview` are the only two options available via `[env.name]`.
```
Why is it [env.name]? I feel like it should be `[env.<ENVIRONMENT>]`? The only available option is the 2 environments listed in the sentence. `name` seems customizable and it isn't.

{{</Aside>}}

> Refer to the following config file for an example of how to override preview deployment configuration:

Reviewer: ```suggestion
Refer to the following `wrangler.toml` configuration file for an example of how to override preview deployment configuration:
```

If you wanted to have configuration values apply to local and preview, but override production, your file would look like this:

```toml
---
filename: wrangler.toml
---
name = "my-pages-site"
pages_build_output_dir = "./dist"

[[kv_namespaces]]
binding = "KV"
id = "<NAMESPACE_ID>"

[vars]
API_KEY = "1234567asdf"

[[env.production.kv_namespaces]]
binding = "KV"
id = "<PRODUCTION_NAMESPACE_ID>"

[[env.production.vars]]
API_KEY = "8901234bfgd"
```

You can always be explicit and override both preview and production:

```toml
---
filename: wrangler.toml
---
name = "my-pages-site"
pages_build_output_dir = "./dist"

[[kv_namespaces]]
binding = "KV"
id = "<NAMESPACE_ID>"

[vars]
API_KEY = "1234567asdf"

[[env.preview.kv_namespaces]]
binding = "KV"
id = "<PREVIEW_NAMESPACE_ID>"

[[env.preview.vars]]
API_KEY = "8901234bfgd"

[[env.production.kv_namespaces]]
binding = "KV"
id = "<PRODUCTION_NAMESPACE_ID>"

[[env.production.vars]]
API_KEY = "6567875fvgt"
```

> ### Important reminder for overrides

Reviewer: Maybe we can just change this to "### Overrides"?

> If you’re using `[env.<NAME>]` to override values, you must specify overrides for all [non-inheritable properties](#non-inheritable-keys) like bindings and variables. It is not possible to override some values and not others.

Reviewer: ```suggestion
If you are using `[env.<NAME>]` to override values, you must specify overrides for all [non-inheritable properties](#non-inheritable-keys) like bindings and variables. It is not possible to override some values and not others.
```
same comment here about `[env.<ENVIRONMENT>]`, `env.name` seems not technically correct in Pages land

  - A date in the form `yyyy-mm-dd`, which will be used to determine which version of the Workers runtime is used. Refer to [Compatibility dates](/workers/configuration/compatibility-dates/).

- `compatibility_flags` {{<type>}}string[]{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - A list of flags that enable features from upcoming features of the Workers runtime, usually used together with `compatibility_date`. Refer to [compatibility dates](/workers/configuration/compatibility-dates/).

- `send_metrics` {{<type>}}boolean{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - Whether Wrangler should send usage metrics to Cloudflare for this project.

- `limits` {{<type-link href="#limits">}}Limits{{</type-link>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - Configures limits to be imposed on execution at runtime. Refer to [Limits](#limits).

- `placement` {{<type-link href="#placement">}}Placement{{</type-link>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - Specify how Pages Functions should be located to minimize round-trip time. Refer to [Smart Placement](/workers/configuration/smart-placement/).

{{</definitions>}}

## Non-inheritable keys

> Non-inheritable keys are configurable at the top-level, but, if any one non-inheritable key is overridden for any environment (ex: `[[env.production.kv_namespaces]]`), all non-inheritable keys must also be specified in the environment config and overridden. Refer to [this section](/pages/functions/wrangler-configuration/#important-reminder-for-overrides) for more details.

Reviewer: ```suggestion
Non-inheritable keys are configurable at the top-level, but, if any one non-inheritable key is overridden for any environment (for example,`[[env.production.kv_namespaces]]`), all non-inheritable keys must also be specified in the environment configuration and overridden. Refer to [Overrides](/pages/functions/wrangler-configuration/#important-reminder-for-overrides) for more details.
```

- `vectorize` {{<type>}}object{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - A list of Vectorize indexes that your Function should be bound to. Refer to [Vectorize indexes](/vectorize/get-started/intro/#3-bind-your-worker-to-your-index).

- `services` {{<type>}}object{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - A list of service bindings that your Function should be bound to. Refer to [service bindings](/pages/functions/bindings/#service-bindings).

- `analytics_engine_datasets` {{<type>}}object{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - Specifies analytics engine datasets that are bound to this Function. Refer to [Workers Analytics Engine](/analytics/analytics-engine/get-started/#2-configure-your-dataset-and-binding-in-wrangler).

- `ai` {{<type>}}object{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - Specifies an AI binding to this Function. Refer to [Workers AI](/pages/functions/bindings/#workers-ai).
{{</definitions>}}

## Limits

You can configure limits for your Pages project in the same way you can for Workers. Read [this guide](/workers/wrangler/configuration/#limits) for more details.

## Bindings

> Bindings are powerful tools that enhance the functionality of your application by connecting it to additional products and resources on the Cloudflare developer platform.

Reviewer: ```suggestion
A [binding](/pages/functions/bindings/) enables your Pages Functions to interact with resources on the Cloudflare Developer Platform. Use bindings to integrate your Pages Functions with Cloudflare resources like [KV](/kv/), [Durable Objects](/durable-objects/), [R2](/r2/), and [D1](/d1/). You can set bindings for both production and preview environments.
```

- Interact with your [environment variables](/pages/functions/bindings/#environment-variables).

### Hyperdrive

[Hyperdrive](/hyperdrive/) bindings allow you to interact with and query any Postgres database from within a Pages Function.

- Configure Hyperdrive bindings via your [`wrangler.toml` file](/workers/wrangler/configuration/#hyperdrive) the same way they are configured with Cloudflare Workers.

### KV namespaces

[Workers KV](/kv/api/) is a global, low-latency, key-value data store. It stores data in a small number of centralized data centers, then caches that data in Cloudflare’s data centers after access.

{{<Aside type="note">}}

When using Wrangler in the default local development mode, files will be written to local storage instead of the preview or production namespace. Refer to [Local development](/workers/testing/local-development/) for more details.

{{</Aside>}}

- Configure KV namespace bindings via your [`wrangler.toml` file](/workers/wrangler/configuration/#kv-namespaces) the same way they are configured with Cloudflare Workers.
- Interact with your [KV namespace binding](/pages/functions/bindings/#kv-namespaces).

### Queues Producers

> [Queues](/queues/) is Cloudflare's global message queueing service, providing [guaranteed delivery](/queues/reference/delivery-guarantees/) and [message batching](/queues/reference/batching-retries/). [Queue Producers](https://developers.cloudflare.com/queues/reference/javascript-apis/#producer) enable you to send messages into a Queue within your Pages Function.

Reviewer: ```suggestion
[Queues](/queues/) is Cloudflare's global message queueing service, providing [guaranteed delivery](/queues/reference/delivery-guarantees/) and [message batching](/queues/reference/batching-retries/). [Queue Producers](/queues/reference/javascript-apis/#producer) enable you to send messages into a queue within your Pages Function.
```

{{<Aside type="note">}}

> Queues Consumers can not currently be bound to Pages Functions

Reviewer: ```suggestion
You cannot currently configure a [queues consumer](/queues/reference/how-queues-works/#consumers) with Pages Functions.
```

- Interact with your [Queues Producer binding](/pages/functions/bindings/#queue-producers).

### R2 buckets

[Cloudflare R2 Storage](/r2) allows developers to store large amounts of unstructured data without the costly egress bandwidth fees associated with typical cloud storage services.

{{<Aside type="note">}}

When using Wrangler in the default local development mode, files will be written to local storage instead of the preview or production bucket. Refer to [Local development](/workers/testing/local-development/) for more details.

{{</Aside>}}

- Configure R2 bucket bindings via your [`wrangler.toml` file](/workers/wrangler/configuration/#r2-buckets) the same way they are configured with Cloudflare Workers.
- Interact with your [R2 bucket bindings](/pages/functions/bindings/#r2-buckets).

### Vectorize indexes

A [Vectorize index](/vectorize/) allows you to insert and query vector embeddings for semantic search, classification and other vector search use-cases.

- Configure Vectorize bindings via your [`wrangler.toml` file](/workers/wrangler/configuration/#vectorize) the same way they are configured with Cloudflare Workers.

### Service bindings

> A service binding allows you to call a Worker from within your Pages Function. This sends HTTP requests to the Worker without those requests going over the Internet. The request immediately invokes the downstream Worker, reducing latency as compared to a request to a third-party service. Refer to [About Service Bindings](/workers/configuration/bindings/about-service-bindings/).

Reviewer: ```suggestion
A service binding allows you to call a Worker from within your Pages Function. Binding a Pages Function to a Worker allows you to send HTTP requests to the Worker without those requests going over the Internet. The request immediately invokes the downstream Worker, reducing latency as compared to a request to a third-party service. Refer to [About Service bindings](/workers/configuration/bindings/about-service-bindings/).
```

- Configure service bindings via your [`wrangler.toml` file](/workers/wrangler/configuration/#service-bindings) the same way they are configured with Cloudflare Workers.
- Interact with your [service bindings](/pages/functions/bindings/#service-bindings).

### Analytics Engine Datasets

[Workers Analytics Engine](/analytics/analytics-engine/) provides analytics, observability and data logging from Pages Functions. Write data points within your Pages Function binding then query the data using the [SQL API](/analytics/analytics-engine/sql-api/).

- Configure Analytics Engine Dataset bindings via your [`wrangler.toml` file](/workers/wrangler/configuration/#analytics-engine-datasets) the same way they are configured with Cloudflare Workers.
- Interact with your [Analytics Engine Dataset](/pages/functions/bindings/#analytics-engine).

### Workers AI

> [Workers AI](/workers-ai/) allows you to run machine learning models, on the Cloudflare network, from your own code –
> whether that be from Workers, Pages, or anywhere via REST API.

Reviewer: ```suggestion
[Workers AI](/workers-ai/) allows you to run machine learning models, on the Cloudflare network, from your own code – whether that be from Workers, Pages, or anywhere via REST API.
```

- Interact with your [Workers AI binding](/pages/functions/bindings/#workers-ai).

## Local development settings

The local development settings that you can configure are the same for Pages Functions and Cloudflare Workers. Read [this guide](/workers/wrangler/configuration/#local-development-settings) for more details.

## Users with existing `wrangler.toml` files

Before you could define your preview and production configuration in `wrangler.toml` — it was possible to use `wrangler.toml` to define which bindings should be available to your Pages project in local development. That means you may already have a file in your project that looks like this:

```toml
---
filename: wrangler.toml
---
[[ kv_namespaces ]]
binding = "KV"
id = "<NAMESPACE_ID>"
```

You can continue to use this file as-is. However when you run `wrangler pages deploy` you will see a warning message telling you that fields are missing and that the file will continue to be used for local development only.

If you would like to use this file for production and preview deployments add the missing fields, update the file as needed, and deploy again.

## Source of truth

> When used in your Pages Functions projects, this file is the source of truth. You will be able to see, but not edit, the same fields when you log into the dashboard.

Reviewer: ```suggestion
When used in your Pages Functions projects, your `wrangler.toml` file is the source of truth. You will be able to see, but not edit, the same fields when you log into the Cloudflare dashboard.
```

---
pcx_content_type: how-to
title: Configuration
weight: 6
---

# Configuration via `wrangler.toml`

Pages Functions can be configured two ways, either via the [Cloudflare dashboard](https://dash.cloudflare.com) or [Wrangler](/workers/wrangler), the Cloudflare Developer Platform CLI. 

This guide will focus on configuring your project via `wrangler.toml`.

> {{<Aside type="note" header="Configuration via `wrangler.toml` is in open beta.">}}

Author: Thanks. Removed.

- **ensure better access control** - by using a configuration file in your repo, you can control who has access to make changes without giving access to your Cloudflare dashboard

## Example `wrangler.toml` file

```toml
---
filename: wrangler.toml
---
name = "my-pages-app"
pages_build_output_dir = "./dist"

[[kv_namespaces]]
binding = "KV"
id = "<NAMESPACE_ID>"

[[d1_databases]]
binding = "DB"
database_name = "northwind-demo"
database_id = "<DATABASE_ID>"

[vars]
API_KEY = "1234567asdf"
```

## Migrating from dashboard configuration

If you have existing Pages projects with configuration setup via the dashboard you can use the following Wrangler command to download your existing configuration and provide you with a valid `wrangler.toml` file:

```sh
$ npx wrangler pages download config <PROJECT_NAME>
```

Run the command, add the file to your project’s root directory, make changes as needed, and create a new deployment either via git integration or direct upload to begin leveraging `wrangler.toml` for your configuration.

> ## Differences between this and Workers

Author: Updated this to be: "Differences using `wrangler.toml` for Pages Functions and Workers"

If you have used [Workers](/workers), you may already be familiar with [`wrangler.toml`](/workers/wrangler/configuration/). There are a few key differences to be aware of when using `wrangler.toml` with your Pages project:

- the configuration fields **do not match exactly** between this file and the Workers equivalent
- the concept of [environments](/pages/functions/wrangler-configuration/#configure-environments) and configuration inheritance in this file **is not** the same as Workers
- this file becomes the [source of truth](/pages/functions/wrangler-configuration/#source-of-truth) when used, meaning that you **can not edit the same fields in the dashboard** once you are using this file

## Configure environments

With `wrangler.toml` you can quickly set configuration across your local environment, preview deployments, and production.

### Local development

> `wrangler.toml` is always available locally via `wrangler pages dev`. This means that you can test out configuration changes quickly without a need to login to the Cloudflare dashboard. Refer to the following config file for an example:

Author: Updated.

[[env.production.kv_namespaces]]
binding = "KV"
id = "<PRODUCTION_NAMESPACE_ID>"

[[env.production.vars]]
API_KEY = "8901234bfgd"
```

You can always be explicit and override both preview and production:

```toml
---
filename: wrangler.toml
---
name = "my-pages-site"
pages_build_output_dir = "./dist"

[[kv_namespaces]]
binding = "KV"
id = "<NAMESPACE_ID>"

[vars]
API_KEY = "1234567asdf"

[[env.preview.kv_namespaces]]
binding = "KV"
id = "<PREVIEW_NAMESPACE_ID>"

[[env.preview.vars]]
API_KEY = "8901234bfgd"

[[env.production.kv_namespaces]]
binding = "KV"
id = "<PRODUCTION_NAMESPACE_ID>"

[[env.production.vars]]
API_KEY = "6567875fvgt"
```

### Important reminder for overrides

If you’re using `[env.<NAME>]` to override values, you must specify overrides for all [non-inheritable properties](#non-inheritable-keys) like bindings and variables. It is not possible to override some values and not others.

This means this configuration will not work:

```toml
---
filename: wrangler.toml
---
name = "my-pages-site"
pages_build_output_dir = "./dist"

[[kv_namespaces]]
binding = "KV"
id = "<NAMESPACE_ID>"

[vars]
API_KEY = "1234567asdf"

[[env.production.vars]]
API_KEY = "8901234bfgd"
```

> This configuration file is missing the production override for `kv_namespaces`, and will fail to validate on deployment.

Author: Added that callout.

  - A date in the form `yyyy-mm-dd`, which will be used to determine which version of the Workers runtime is used. Refer to [Compatibility dates](/workers/configuration/compatibility-dates/).

- `compatibility_flags` {{<type>}}string[]{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - A list of flags that enable features from upcoming features of the Workers runtime, usually used together with `compatibility_date`. Refer to [compatibility dates](/workers/configuration/compatibility-dates/).

- `send_metrics` {{<type>}}boolean{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - Whether Wrangler should send usage metrics to Cloudflare for this project.

- `limits` {{<type-link href="#limits">}}Limits{{</type-link>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - Configures limits to be imposed on execution at runtime. Refer to [Limits](#limits).

- `placement` {{<type-link href="#placement">}}Placement{{</type-link>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - Specify how Pages Functions should be located to minimize round-trip time. Refer to [Smart Placement](/workers/configuration/smart-placement/).

{{</definitions>}}

## Non-inheritable keys

> Non-inheritable keys are configurable at the top-level, but, if an environment override is present, these must all be specified in the environment config and overridden.

Author: Added this. This is a good example!

  - A list of R2 buckets that your Function should be bound to. Refer to [R2 buckets](/pages/functions/bindings/#r2-buckets).

- `vectorize` {{<type>}}object{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - A list of Vectorize indexes that your Function should be bound to. Refer to [Vectorize indexes](/vectorize/get-started/intro/#3-bind-your-worker-to-your-index).

- `services` {{<type>}}object{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - A list of service bindings that your Function should be bound to. Refer to [service bindings](/pages/functions/bindings/#service-bindings).

- `analytics_engine_datasets` {{<type>}}object{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - Specifies analytics engine datasets that are bound to this Function. Refer to [Workers Analytics Engine](/analytics/analytics-engine/get-started/#2-configure-your-dataset-and-binding-in-wrangler).

- `ai` {{<type>}}object{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - Specifies an AI binding to this Function. Refer to [Workers AI](/pages/functions/bindings/#workers-ai).
{{</definitions>}}

## Limits

> Limits can be configured the same way they are for Workers. Read [this guide](/workers/wrangler/configuration/#limits) for more details.

Author: Updated.

## Bindings

Bindings are powerful tools that enhance the functionality of your application by connecting it to additional products and resources on the Cloudflare developer platform.

### D1 databases

[D1](/d1/) is Cloudflare's serverless SQL database. A Function can query a D1 database (or databases) by creating a [binding](/workers/configuration/bindings/) to each database for D1's [client API](/d1/build-with-d1/d1-client-api/).

{{<Aside type="note">}}

When using Wrangler in the default local development mode, files will be written to local storage instead of the preview or production database. Refer to [Local development](/workers/testing/local-development/) for more details.

{{</Aside>}}

> - Configure D1 database bindings via your [`wrangler.toml` file](/workers/wrangler/configuration/#d1-databases) the same way they are configured with Cloudflare Workers.

Author: It is, and I could. I thought this could match the content we already have in other places like https://developers.cloudflare.com/workers/configuration/bindings/. My concern with only having it in one place is that a person looking for a specific binding could miss it.

> - Configure D1 database bindings via your [`wrangler.toml` file](/workers/wrangler/configuration/#d1-databases) the same way they are configured with Cloudflare Workers.

Reviewer: Partials as a longer term project here.

- Configure Analytics Engine Dataset bindings via your [`wrangler.toml` file](/workers/wrangler/configuration/#analytics-engine-datasets) the same way they are configured with Cloudflare Workers.
- Interact with your [Analytics Engine Dataset](/pages/functions/bindings/#analytics-engine).

### Workers AI

[Workers AI](/workers-ai/) allows you to run machine learning models, on the Cloudflare network, from your own code – whether that be from Workers, Pages, or anywhere via REST API.

{{<render file="_ai-local-usage-charges.md" productFolder="workers">}}

Unlike other bindings, this binding is limited to one AI binding per Pages Function project.

- Configure Workers AI bindings via your [`wrangler.toml` file](/workers/wrangler/configuration/#workers-ai) the same way they are configured with Cloudflare Workers.
- Interact with your [Workers AI binding](/pages/functions/bindings/#workers-ai).

## Local development settings

The local development settings that you can configure are the same for Pages Functions and Cloudflare Workers. Read [this guide](/workers/wrangler/configuration/#local-development-settings) for more details.

## Users with existing `wrangler.toml` files

> Before you could define your preview and production configuration in `wrangler.toml` — it was possible to use `wrangler.toml` to define which bindings should be available to your Pages project in local development. That means you may already have a file in your project that looks like this:

Reviewer: add the pages-build-output-dir key for your wrangler.toml to be used in deployments

---
pcx_content_type: how-to
title: Wrangler Configuration
weight: 6

> ---

Author: I've updated this to be "Configuration" within the Functions section. `wrangler.toml` is only available to full-stack projects using Functions that's why I put it within that section. 

As to the slug, I'm open to other ideas, but am good with it as-is.

# Configure `wrangler.toml`

> Pages Functions can be configured via a `wrangler.toml` file in your project's root directory. `wrangler.toml` allows you to configure things like: bindings, environment variables, development settings, and runtime settings alongside the rest of your code. 

Author: I've updated this with links to what is being referred to.

Using this file allows you to:

- **store your configuration file in source control** - keep your configuration in your repo alongside the rest of your code
- **edit your config via your code editor** - remove the need to switch back and forth between interfaces
- **write configuration that is shared across environments** - define config like bindings and environment variables for local, preview, and production in one file
- **ensure better access control** - by using a configuration file in your repo, you can control who has access to make changes without giving access to your Cloudflare dashboard

> If you have used Workers in the past, you may already be familiar with [this file](/workers/wrangler/configuration/). It should be noted that there are differences with the way this file behaves for Pages Functions, and that the configuration fields do not match exactly between this file and the Workers equivalent.

Author: I've added a section to point out these differences.

## Example `wrangler.toml` file

```toml
---
filename: wrangler.toml
---
name = "my-pages-app"
pages_build_output_dir = “./dist”

[[ kv_namespaces ]]
binding = "KV"
id = "<NAMESPACE_ID>"

[[ d1_databases ]]
binding = "DB"
database_name = "northwind-demo"
database_id = "<DATABASE_ID>"

[vars]
API_KEY = "1234567asdf"
```

> 

Reviewer 2: Feels like somewhere early on in this page there should be a quick way for me to understand at-a-glance, which parts of wrangler.toml are supported in Pages just like Workers, which are unique to pages (ex: build output directory) — kind of like compat table. Right now a lot to sift through

Author: I wanted to keep the structure of this page similar to what we have at https://developers.cloudflare.com/workers/wrangler/configuration/

Maybe we add a compat table at some point, but I think we show users what's possible and don't assume (at this point) the user reading this has a concept of Workers yet.

## Environments

> With `wrangler.toml` you can quickly set default configuration across your local environment, preview deployments, and production.

Author: I think this should stay in. Why are you suggesting to remove it?

### Local development

Top-level config is always available locally via `wrangler pages dev`, meaning that you can test out config changes quickly without going to the Cloudflare dashboard. For example, look at the following config file:

```toml
---
filename: wrangler.toml
---
name = "my-pages-app"
pages_build_output_dir = “./dist”
compatibility_date = "2023-10-12"
compatibility_flags = [“nodejs_compat”]

[[ kv_namespaces ]]
binding = "KV"
id = "<NAMESPACE_ID>"
```

> This adds compatibility settings like `nodejs_compat` and a KV namespace binding to the project. Once saved, run `wrangler pages dev` and the compatibility settings will apply locally. You can access the KV binding in your Pages Function code at `context.env.KV`.

Author: Updated this to be clearer.

{{<Aside type="note">}}

For a full list of configuration keys read about [inheritable keys](#inheritable-keys) and [non-inheritable keys](#non-inheritable-keys).

{{</Aside>}}

### Production and Preview Deployments

Once you are ready to deploy your project, you can set the configuration for production and preview deployments by creating a new deployment containing a `wrangler.toml` file. This can be the same configuration you use locally or you can specify overrides.

{{<Aside type="note">}}

> For the following commands, it is important to remember the branch that you set as your [production branch](/pages/configuration/branch-build-controls/#production-branch-control) as well as your [preview branch settings](/pages/configuration/branch-build-controls/#preview-branch-control).

Reviewer: @admah I am confused as to the order of events here. Is the user defining prod and preview branch as part of `wrangler.toml` setup? Reading the docs, defining prod and preview branches seems like a process that's done **only** in the dashboard actually. in which case, this wouldn't need to exist on this page because it's a deployment detail and not a wrangler.toml necessity?

`--branch` is also an optional option on deploy command, I'm not sure how this section this connects to wrangler.toml.

{{</Aside>}}

To use the configuration example above for production configuration, you can run:

```sh
$ npx wrangler pages deploy --branch <PRODUCTION BRANCH>
```

If you haven't set `pages_build_output_dir` locally, you will need to run:

```sh
$ npx wrangler pages deploy <BUILD_OUTPUT_DIRECTORY> --branch <PRODUCTION BRANCH>

> ```

Author: I've removed it 👍

[env.preview.vars]
API_KEY = "8901234bfgd"
```

If you deployed this file via `wrangler pages deploy`, the top-level configuration would apply to local and production, while `env.preview` would override the configuration for preview deployments.

If you wanted to have top-level configuration apply to local and preview, but override production, your file would look like this:

```toml
---
filename: wrangler.toml
---
name = "my-pages-site"
pages_build_output_dir = “./dist”

[[ kv_namespaces ]]
binding = "KV"
id = "<NAMESPACE_ID>"

[vars]
API_KEY = "1234567asdf"

[env.production.kv_namespaces]
binding = "KV"
id = "<PRODUCTION_NAMESPACE_ID>"

[env.production.vars]
API_KEY = "8901234bfgd"
```

You can always be explicit and override preview and production:

```toml
---
filename: wrangler.toml
---
name = "my-pages-site"
pages_build_output_dir = “./dist”

[[ kv_namespaces ]]
binding = "KV"
id = "<NAMESPACE_ID>"

[vars]
API_KEY = "1234567asdf"

[env.preview.kv_namespaces]
binding = "KV"
id = "<PREVIEW_NAMESPACE_ID>"

[env.preview.vars]
API_KEY = "8901234bfgd"

[env.production.kv_namespaces]
binding = "KV"
id = "<PRODUCTION_NAMESPACE_ID>"

[env.production.vars]
API_KEY = "6567875fvgt"
```

> ### Important reminder for overrides

Reviewer 2: Can be more specific?

- `r2_buckets` {{<type>}}object{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - A list of R2 buckets that your Function should be bound to. Refer to [R2 buckets](/pages/functions/bindings/#r2-buckets).

- `vectorize` {{<type>}}object{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - A list of Vectorize indexes that your Function should be bound to. Refer to [Vectorize indexes](/vectorize/get-started/intro/#3-bind-your-worker-to-your-index).

- `services` {{<type>}}object{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - A list of service bindings that your Function should be bound to. Refer to [service bindings](/pages/functions/bindings/#service-bindings).

- `analytics_engine_datasets` {{<type>}}object{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - Specifies analytics engine datasets that are bound to this Function. Refer to [Workers Analytics Engine](/analytics/analytics-engine/get-started/#2-configure-your-dataset-and-binding-in-wrangler).

- `ai` {{<type>}}object{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - Specifies an AI binding to this Function. Refer to [Workers AI](/workers-ai/configuration/bindings/).

{{</definitions>}}

> ## Limits

Reviewer 2: This section is same as https://developers.cloudflare.com/workers/wrangler/configuration/#limits — link to there and say that it works the same way with Pages Functions?

---
mtls_certificates = [
    { binding = "<BINDING_NAME1>", certificate_id = "<CERTIFICATE_ID1>" },
    { binding = "<BINDING_NAME2>", certificate_id = "<CERTIFICATE_ID2>" }
]

# or

[[mtls_certificates]]
binding = "<BINDING_NAME1>"
certificate_id = "<CERTIFICATE_ID1>"

[[mtls_certificates]]
binding = "<BINDING_NAME2>"
certificate_id = "<CERTIFICATE_ID2>"
```

mTLS certificate bindings can then be used at runtime to communicate with secured origins via their [`fetch` method](/workers/runtime-apis/bindings/mtls).

### Workers AI

[Workers AI](/workers-ai/) allows you to run machine learning models, on the Cloudflare network, from your own code –
whether that be from Workers, Pages, or anywhere via REST API.

{{<render file="_ai-local-usage-charges.md" productFolder="workers">}}

Unlike other bindings, this binding is limited to one AI binding per Pages Function project.

{{<definitions>}}

- `binding` {{<type>}}string{{</type>}} {{<prop-meta>}}required{{</prop-meta>}}

  - The binding name.

{{</definitions>}}

Example:

```toml
---
filename: wrangler.toml
---
ai = { binding = "<AI>" }

# or

[ai]
binding = "AI" # available in your Pages Functions code on `context.env.AI`
```

> ## Local development settings

Reviewer 2: Is any of this different for Pages vs. Workers? If not, can be clear, link to right place?

Author: Updated and linked to Workers docs.

  - IP address for the local dev server to listen on. Defaults to `localhost`.

- `port` {{<type>}}number{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  - Port for the local dev server to listen on. Defaults to `8787`.

- `local_protocol` {{<type>}}string{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  -  Protocol that local dev server listens to requests on. Defaults to `http`.

- `upstream_protocol` {{<type>}}string{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  -  Protocol that the local dev server forwards requests on. Defaults to `https`.

- `host` {{<type>}}string{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

  -  Host to forward requests to, defaults to the host of the first `route` of the Worker.

{{</definitions>}}

Example:

```toml
---
header: wrangler.toml
---
[dev]
ip = "192.168.1.1"
port = 8080
local_protocol = "http"
```

## Migrating from dashboard configuration

If you have existing Pages projects with configuration setup via the dashboard we provide a Wrangler command that will download your existing configuration and provide you with a valid `wrangler.toml` file.

```sh

> $ npx wrangler pages download config <PROJECT_NAME>

Reviewer 2: This feels like important feature but is way, way at the bottom?

Author: Moved to higher for better visibility.

```

Run the command, add the file to your project’s root directory, make changes as needed, and deploy to begin leveraging `wrangler.toml` in your project.

## Users with existing `wrangler.toml` files

It has been possible to use `wrangler.toml` for development with local bindings. That means you may have a file in your project that looks like this:

```toml
---
filename: wrangler.toml
---
[[ kv_namespaces ]]
binding = "KV"
id = "<NAMESPACE_ID>"
```

You can continue to use this file as-is. However when you run `wrangler pages deploy` you will see a warning message telling you that fields are missing and that the file will continue to be used for local development only.

> If you would like to use this file for production and preview deployments add the missing fields, update the file as needed, and deploy again.

Reviewer 2: This doesn't _quite_ explain what will happen in base case, in CI/CD, "will my existing `wrangler.toml` suddenly define what gets deployed?" is reasonable question for someone to ask.
